File: 3Sum-Closest/3Sum-Closest.py


class Solution:
    # @return an integer
    def threeSumClosest(self, num, target):
        num.sort()
        first=True
        for i in range(len(num)):
            start=i+1
            end=len(num)-1
            while start<end:
                temp=num[i]+num[start]+num[end]
                if first:
                    ans=temp
                    first=False
                if abs(temp-target)<abs(ans-target):
                    ans=temp
                if target==temp:
                    return ans
                elif temp<target:
                    start+=1
                else:
                    end-=1
        return ans
 

File: 3Sum-Closest/3Sum-Closest.cpp

class Solution {
public:
    int threeSumClosest(vector<int> &num, int target) {
        sort(num.begin(),num.end());
        int n=num.size();
        int ans=0;
        bool first=true;
        for(int i=0;i<n;i++)
        {
            int start=i+1;
            int end=n-1;
            while(start<end)
            {
                int temp=num[i]+num[start]+num[end];
                if(temp<target)
                    start++;
                else if(temp==target)
                    return temp;
                else
                    end--;
                if(first || abs(temp-target)<abs(ans-target))
                {
                    ans=temp;
                    first=false;
                }
            }
        }
        return ans;
    }
};


 